src/main/java/appeng/crafting/v2/resolvers/CraftableItemResolver.java

```diff
@@ -217,37 +228,80 @@ public StepOutput calculateOneStep(CraftingContext context) {
+                    context.byproductsInventory.injectItems(
+                            matchingOutput.copy().setStackSize(newRemainder - oldRemainder),
+                            Actionable.MODULATE,
+                            context.actionSource);
+                } else {
+                    context.byproductsInventory.extractItems(
+                            matchingOutput.copy().setStackSize(oldRemainder - newRemainder),
+                            Actionable.MODULATE,
+                            context.actionSource);
+                }
+                this.matchingOutputRemainderItems = newRemainder;
             }
-            for (Entry<IAEItemStack, Long> entry : byproducts.entrySet()) {
-                final IAEItemStack byproductStack = entry.getKey();
-                final long perCraft = entry.getValue();
-                context.itemModel.extractItems(
-                        byproductStack.copy().setStackSize(perCraft * amount),
-                        Actionable.MODULATE,
-                        context.actionSource);
-                entry.getKey().setStackSize(byproductStack.getStackSize() - amount * perCraft);
+            if (newTotalCrafts != oldTotalCrafts) {
+                if (newTotalCrafts > oldTotalCrafts) {
+                    throw new IllegalStateException(
+                            "Refund total crafts invariant broken: " + newTotalCrafts + " - " + this);
+                }
+                this.totalCraftsDone = newTotalCrafts;
+                final long craftsRefunded = oldTotalCrafts - newTotalCrafts;
+                for (RequestAndPerCraftAmount subrequest : childRequests.values()) {
+                    subrequest.request.partialRefund(context, subrequest.perCraftAmount * craftsRefunded);
+                }
+                for (Entry<IAEItemStack, Long> entry : byproducts.entrySet()) {
+                    final IAEItemStack byproductStack = entry.getKey();
+                    final long perCraft = entry.getValue();
+                    context.byproductsInventory.extractItems(
+                            byproductStack.copy().setStackSize(perCraft * craftsRefunded),
+                            Actionable.MODULATE,
+                            context.actionSource);
+                    entry.getKey().setStackSize(byproductStack.getStackSize() - craftsRefunded * perCraft);
+                }
             }
+            this.fulfilledAmount = newFulfilled;
+            return oldFulfilled - newFulfilled;
         }
 
         @Override
         public void fullRefund(CraftingContext context) {
             totalCraftsDone = 0;
+            fulfilledAmount = 0;
             childRequests.values().forEach(req -> req.request.fullRefund(context));
             childRequests.clear();
             childRecursionRequests.values().forEach(req -> req.fullRefund(context));
             childRecursionRequests.clear();
+            // extract all byproducts because they are no longer produced
```

Comment:
Why it is necessary to extract everything to clear the byproducts, wouldn't `clear` do that anyway?

Reply:
This removes them from the modelled crafting cpu inventory - `clear` just clear the list of stacks internal to this task. It's the backtracking step in this algorithm - if something that requested this craft no longer needs as many items (e.g. because it's a crafting task that couldn't be fully satisfied), partialRefund should undo all the changes resolving this task did to leave it as-if it was recalculated with a smaller initial request. This means that because originally byproducts were inserted into the modelled inventory, the extras have to be extracted during backtracking (but only the ones inserted by this task, because byproducts from other tasks still remain there).